### ISSUE TITLE
Sensible Package Manager Defaults For Auto-Install

### ISSUE THREAD

**Issue by zor-el** [parcel-bundler/parcel#3919]
This proposal is an attempt to be equally inclusive to both npm and yarn users with regard to an unclouded zero configuration experience.

### TL;DR

**In the absence of a lock file, Parcel should default to the package manager that was used to install it.**

In other words: If Parcel was installed via `npm`, `npm` should be used. If it was installed via `yarn`, `yarn` should be used.

That way, both npm users and yarn users will have their preferred zero configuration experience with Parcel as long as they use their favorite package manager to install it.

### Background

Before #2119 Parcel used a mostly undiscriminating differentiator between npm users and yarn users to select a default package manager in the absence of a lock file. The old policy aimed at making the most likely preferred choice based on what is installed on the user's system, following this logic (a recap from [my comment on #2117](https://github.com/parcel-bundler/parcel/issues/2117#issuecomment-565807685)):

> `npm` is not a choice. It comes bundled with Node. It's presence is not a differentiating factor between npm users and yarn users. The presence of `yarn` is, as its installation is a conscious choice.

Alas, the vocality of discussions against this policy (ex.  #1923, #2117) seem to have drowned the rationality for retaining it, and a premature PR found its way into Parcel (#2119). That PR was scheduled for reversal (#2125), but it [hasn't been merged b/c it would break SemVer](https://github.com/parcel-bundler/parcel/issues/2117#issuecomment-565811741). (Interestingly enough, #2125 was a breaking change, too, but got past SemVer restrictions b/c it was flagged as a bug, when it should have probably better been an RFC.)

### Rationale

PR #2119 was accepted with "npm is the de-facto standard package manager" as essentially the only argument AFAICT.

But: **It's not about package managers. It's about people.**

However, the mere presence of `yarn` is not an ideal differentiator either, as it might not be an expression of preference but the result of necessity (ex. b/c a single project required it). In that case npm'ers would be treated by Parcel as yarn'ers even if they aren't. (The current situation is worse though, b/c it treats all yarn'ers as npm'ers no matter what.)

I am suggesting that the package manager that has actually been used to install Parcel might be a better indicator of preference, as it is reasonable to assume that a yarn'er would use `yarn` to install Parcel rather than `npm` and vice versa.

**Comment by mischnic** [comment]
> In the absence of a lock file

Are you referring to the case where Parcel is installed locally and `npm --no-package-lock` was used, or when Parcel was installed globally?


(The relevant logic for Parcel 2:)
https://github.com/parcel-bundler/parcel/blob/e2ea84373758ef22da979bc7abaf9b8ca29a1f59/packages/core/package-manager/src/installPackage.js#L74-L91

**Comment by zor-el** [comment]
> Are you referring to the case where Parcel is installed locally and `npm --no-package-lock` was used, or when Parcel was installed globally?

When parcel was installed globally (which IMJ is the primary use case).

But the other case is an interesting one, too. In that case, an analogously sensible strategy might be to take the invocation mechanism into account. I.e. invocation via `npx parcel` would suggest that particular project requires the use of npm (regardless of the user's general preference). (I can see though how such edge cases would complicate an implementation.)

**Comment by mischnic** [comment]
> When parcel was installed globally (which is the primary use case with parcel AFAICT).

We recommend not doing that (and I don't know how big the share of global installations is), because it isn't as easy to upgrade.

And using the package manager that globally installed Parcel for the project autoinstall would only make a difference for the first package?

**Comment by zor-el** [comment]
> We recommend not doing that [install parcel globally]

That is still the officially recommended way though. Here is a short recap of Parcels current [Getting Started](https://parceljs.org/getting_started.html):

```sh
yarn global add parcel-bundler # or npm install -g parcel-bundler
yarn init -y # or npm init -y
# create index.html
#   reference files in various formats from in there
#   ex. <link rel="stylesheet" href="app.sass">
parcel index.html # takes care of all setup and transforms under the hood
```

With the *current release* this would result in parcel using `npm` regardless of a presence of yarn. Before #2119 it would have defaulted to yarn if present, but that has annoyed npm users who happened to have yarn installed out of necessity only. (Hence this RFC. I use yarn, but I understand the annoyance.)

> (and I don't know how big the share of global installations is),

I'd guess quite big if we were to assume people followed the Getting Started instructions. (And I don't see why they wouldn't, given Parcels' zero-configuration-no-worries USP.)

> because it isn't as easy to upgrade.

How so?

> And using the package manager that globally installed Parcel for the project autoinstall would only make a difference for the first package?

How so?

**Comment by mischnic** [comment]
> That is still the officially recommended way though. Here is a short recap of Parcels Getting Started at time of this writing:

😬 

> > And using the package manager that globally installed Parcel for the project autoinstall would only make a difference for the first package?
>
> How so?

After the first package was auto installed, there would be a lockfile and the package manager will be determined based on that lockfile.

>> because it isn't as easy to upgrade.
>
> How so?

Suppose you want to use two different versions of Parcel in two project, that isn't possible with a global installation.

**Comment by zor-el** [comment]
> Suppose you want to use two different versions of Parcel in two project, that isn't possible with a global installation.

My take: Parcel is a tool, and such tool isolation would only be needed if parcel-bundled app code would behave differently depending on the version of parcel used to bundle it. Performance variances aside, that would be scary. (I like to assume compilers/linkers/bundlers are semantics-preserving.)

I understand there is a place for locally isolated tooling, but IMO a globally available parcel is closer to Parcels original value proposition, exemplified exactly in the four simple steps of the original Getting Started document. I truly hope Parcel 2 does not introduce a departure from this wonderfully unique concept.
